File: frontend/libtrack-ui/src/app/books/book-mapper.ts
import {Book} from './book';

export function formToBook(formValue: any): Partial<Book> {
  const book: Partial<Book> = {};
  book.title  = formValue.title;
  book.author = formValue.author;
  book.genre  = formValue.genre;

  book.isbn            = Number(formValue.isbn);
  book.totalCopies     = Number(formValue.totalCopies);
  book.availableCopies = Number(formValue.availableCopies);

  return book;
  }

export function formToPartialBook(formValue: any): Partial<Book> {
  const book: Partial<Book> = {};
  if (formValue.title)  book.title  = formValue.title;
  if (formValue.author) book.author = formValue.author;
  if (formValue.genre)  book.genre  = formValue.genre;
  if (formValue.isbn !== '' && formValue.isbn !== null)
    book.isbn = Number(formValue.isbn);
  if (formValue.totalCopies !== '' && formValue.totalCopies !== null)
    book.totalCopies = Number(formValue.totalCopies);
  if (formValue.availableCopies !== '' && formValue.availableCopies !== null)
    book.availableCopies = Number(formValue.availableCopies);
  return book;
}

File: backend/src/main/java/com/berk/libtrack/mappers/BookMapper.java
package com.berk.libtrack.mappers;

import com.berk.libtrack.domain.dto.BookDto;
import com.berk.libtrack.domain.entities.BookEntity;
import org.springframework.stereotype.Component;

@Component
public class BookMapper implements Mapper<BookEntity, BookDto>{


    @Override
    public BookDto mapTo(BookEntity bookEntity) {
        return BookDto.builder()
                .id(bookEntity.getId())
                .isbn(bookEntity.getIsbn())
                .title(bookEntity.getTitle())
                .author(bookEntity.getAuthor())
                .genre(bookEntity.getGenre())
                .totalCopies(bookEntity.getTotalCopies())
                .availableCopies(bookEntity.getAvailableCopies())
                .createdAt(bookEntity.getCreatedAt())
                .build();
    }

    @Override
    public BookEntity mapFrom(BookDto bookDto) {
        return BookEntity.builder()
                .id(bookDto.getId())
                .isbn(bookDto.getIsbn())
                .title(bookDto.getTitle())
                .author(bookDto.getAuthor())
                .genre(bookDto.getGenre())
                .totalCopies(bookDto.getTotalCopies())
                .availableCopies(bookDto.getAvailableCopies())
                .createdAt(bookDto.getCreatedAt())
                .build();
    }
}
